pkg/provider/pingfed/pingfed.go

```diff
@@ -226,10 +477,48 @@ func extractSAMLResponse(doc *goquery.Document) (v string, ok bool) {
 	return doc.Find("input[name=\"SAMLResponse\"]").Attr("value")
 }
 
+func docIsDuoIFrame(doc *goquery.Document) bool {
+	return doc.Find("title").Text() == "Duo Security Two-Factor Authentication"
+}
+
+func docIsDuoTwoFactor(doc *goquery.Document) bool {
+	return doc.Find("title").Text() == "\nTwo-Factor Authentication\n"
+}
+
+func docIsDuoStat(doc *goquery.Document) bool {
+	htmlResp, _ := doc.Selection.Html()
+	htmlResp = html.UnescapeString(htmlResp)
+	return strings.Contains(htmlResp, "stat") && strings.Contains(htmlResp, "txid")
+}
+
+func docIsDuoPush(doc *goquery.Document) bool {
+	htmlResp, _ := doc.Selection.Html()
+	htmlResp = html.UnescapeString(htmlResp)
+	return gjson.Get(htmlResp, "response.status_code").String() == "pushed"
+}
+
+func docIsDuoCookie(doc *goquery.Document) bool {
+	htmlResp, _ := doc.Selection.Html()
+	htmlResp = html.UnescapeString(htmlResp)
+	return strings.Contains(htmlResp, "cookie") && strings.Contains(htmlResp, "stat")
+}
+
 // ensures given url is an absolute URL. if not, it will be combined with the base URL
 func makeAbsoluteURL(v string, base string) string {
 	if u, err := url.ParseRequestURI(v); err == nil && !u.IsAbs() {
 		return fmt.Sprintf("%s%s", base, v)
 	}
 	return v
 }
+
+func parseTokens(blob string) (string, string, string, string) {
+
+	hostRgx := regexp.MustCompile(`'host': '(.*?)'`)
+	sigRgx := regexp.MustCompile(`'sig_request': '(.*?)'`)
+	dpaRgx := regexp.MustCompile(`'post_action': '(.*?)'`)
+	dataSigRequest := sigRgx.FindStringSubmatch(blob)
+	duoHost := hostRgx.FindStringSubmatch(blob)
+	postAction := dpaRgx.FindStringSubmatch(blob)
+	duoSignatures := strings.Split(dataSigRequest[1], ":")
```

Comment:
I would check the array has assumed number of values before accessing an index.

Reply:
Added check for assumed length before accessing